compiler/noirc_frontend/src/hir/comptime/interpreter.rs

```diff
@@ -665,13 +664,10 @@ impl<'local, 'interner> Interpreter<'local, 'interner> {
                         } else {
                             result.push_str(&value.display(self.elaborator.interner).to_string());
                         }
+                    } else {
+                        // TODO: should we panic here?
```

Comment:
TODO: should we panic if we are interpreting a fmtstr, get into an interpolation but we can't find the value to interpolate?

Reply:
I see that you one expression per fragment [here](https://github.com/noir-lang/noir/pull/6703/files#diff-3c108dc343290b597e508f0b0cc501f5a848f3a644aa08a406a9bfb1a2305f73R273), but if one of the fragments was undeclared this is [skipped](https://github.com/noir-lang/noir/pull/6703/files#diff-3c108dc343290b597e508f0b0cc501f5a848f3a644aa08a406a9bfb1a2305f73R264), so the length of them can differ. I'm not sure such a string will be interpreted, but if it can, we shouldn't panic.

Reply:
OTOH maybe the two independent vectors are not the right way to represent a `HirLiteral::FmtString`; it could be a single `Vec<(FmtStrFragment, Option<ExprId>)>` instead, because currently it would conceptually cause fragments to be interpolated into the wrong spot if one is missing.

Reply:
That makes sense! I'll do this change.

Reply:
Hm, or maybe not... if we can't type-check an interpolation we could simply push the fragment as-is without interpolation. I don't know if that part of the code will be reached, but just in case... all of this probably doesn't matter much as we are already in an error state 🤷

Reply:
yeah so currently as I understand it if we have, say `f"foo={foo}, bar={non_existing}, baz={baz}"`, then it would become something like 
```rust
HirLiteral::FmtStr {
fragments: vec![
  S("foo="), 
  I("{foo}"), 
  S(", bar="), 
  I("{non_existing}"), 
  S(", baz="), 
  I("{baz}")
], 
captures:  vec![
  foo_id, 
  baz_id
]}
```
and then you go over the `fragments` and if it's an `S` you push as-is, if it's an `I` then it pops the next value from `captures`. This will result in: `"foo={eval(foo_id)}, bar={eval(baz_id)}, baz="`. 

What I'm suggesting is to represent it like this:
```rust
HirLiteral::FmtStr(
vec![
  S("foo="), 
  I("{foo}", Some(foo_id)), 
  S(", bar="), 
  I("{non_existing}", None), 
  S(", baz="), 
  I("{baz}", Somę(baz_id))
])
```

And then as you go through them, you see that the 2nd `I` has no corresponding expression to evaluate, so you can display it as `"foo={eval(foo_id)}, bar={{non_existing}}, baz={eval(baz_id)}"` or `bar=, baz={eval(...` or whatever you fancy.

Reply:
I like this! I'll do it.

Reply:
Hm, I think for this I'll need to introduce a new type that's that `S` or `I` but where `I` has the string fragment plus `Option<ExprId>`. That makes the code a bit longer and complex, but I'm still thinking this is for an error case that won't go to the monomorphizer, etc., so handling all of this with extra logic might not be worth it.

Reply:
Another way of looking at it is that the interpreter already does handle it, but in a weird way where captures appear in the wrong position 🤷 So we have extra logic that might be illogical :D

Reply:
I think this would be good to test for to see if we can actually trigger this behavior in real comptime code. The variable not declared error comes from elaborating the format string, so it should be sufficient to test if we can do this by calling a comptime function with a format string in it.

As far as the original TODO goes, we should definitely error. If there was an error evaluating the string we should return an error instead of a partial string or something. We may still need the partial string approach @aakoshh is talking about for the elaborator though depending on if the error mentioned above stops the interpreter from interpreting the bad string or not though.

Reply:
Huh, interesting. This code:

```noir
fn main() {
    comptime {
        let x = 1;
        let s = f"hello {x} {y}";
        println(s);
    }
}
```

gives an error, but before erroring it prints this:

```
hello 1
```

If we change it to `f"hello {y} {x}" it prints the same thing, because it ignores which one goes where (as you noticed).

But... I still don't know if it's worth doing this more correctly. It's also not clear what should be printed here (do we print "{y}" or do we show "ERROR" or somehting like that). Also, this code was working like this before this PR, so...

Maybe others from @noir-lang/core could give their opinion here.

Reply:
As usual, I wrote the comment before seeing Jake replied...

Reply:
Okay, so when you `execute` the code above it gets compiled first, and during compilation the interpreter evaluates the code and prints the incorrect string. Then the compiler prints the errors from the elaborator, and finally the `execute` command errors because there were errors. 

If we panicked in the interpreter the entire program would crash before the user could see what was wrong:

```console
❯ cargo run -q -p nargo_cli -- --program-dir . execute
The application panicked (crashed).
Message:  oh noes!
Location: compiler/noirc_frontend/src/hir/comptime/interpreter.rs:678

This is a bug. We may have already fixed this in newer versions of Nargo so try searching for similar issues at https://github.com/noir-lang/noir/issues/.
If there isn't an open issue for this bug, consider opening one at https://github.com/noir-lang/noir/issues/new?labels=bug&template=bug_report.yml
```

I'd say printing the correct string would still be helpful, otherwise it might confuse come macro developer. Although the error is pretty clear so it probably doesn't matter that much:

```console
❯ cargo run -q -p nargo_cli -- --program-dir . execute
hello y=1 x=
error: cannot find `y` in this scope
  ┌─ src/main.nr:4:17
  │
4 │         let s = f"hello y={y} x={x}";
  │                 -------------------- not found in this scope
  │

Aborting due to 1 previous error
```

I tried returning an `Err` from `evaluate_format_string` but it did not improve the above, mostly because I used some random `InterpreterError`. I suppose one that would stay "cannot evaluate format string" could be added. It depends on whether we think there is a point in trying to carry on with the _rest_ of the program after producing a partially invalid string. For println it's probably ok, but if they use it for something it might just cause further problems.

Reply:
> If we panicked in the interpreter the entire program would crash before the user could see what was wrong:
> ...
> I tried returning an Err from evaluate_format_string but it did not improve the above, mostly because I used some random InterpreterError.

We should definitely return an Err when evaluating a bad format string in the interpreter. It's what we do for other errors, even ones that could already be reported by the elaborator, such as type errors. Worst case we may get 2 similar errors printed for the same issue but it is worth it to stop code after the format string from being executed.

This is very forced but imagine a code with side-effects at compile-time:

```
comptime fn foo(name, balance, other_number) {
    let s = f"{{ name: {name}, balance: {balance}, other: {other_number} }}";
    process_transaction(s);
}
```
If this gets formatted using `other_number` in place of `balance` that could be bad™️ for users. Not that they should use format strings in their transaction logic either way.

Reply:
Makes sense. Then I'll change the code to how Akosh suggested it.

Reply:
Actually, I'll just return `Err` in the case where we can't find a value to interpolate (easier).

Reply:
I agree, this is what I meant by
> but if they use it for something it might just cause further problems.

Reply:
Done!

Reply:
We can't do `fragments.len() != captures.len()` because `fragments` also has both the raw string values and the pieces we need to interpolate to.

Reply:
Right, you would have to count the `Interpolate` types. I guess the only difference would be that this way it might error during the evaluation of one of the arguments, before you run out of captures.